text/0258-env-context-baggage-carriers.md

```diff
@@ -0,0 +1,236 @@
+# Environment Variable Specification for Context and Baggage Propagation
+
+This is a proposal to add Environment Variables to the OpenTelemetry
+specification as a carrier for context and baggage propagation between
+processes.
+
+## Table of Contents
+
+* [Motivation](#motivation)
+* [Explanation](#explanation)
+* [Internal Details](#internal-details)
+  * [UNIX](#unix)
+  * [Windows](#windows)
+  * [Allowed Characters](#allowed-characters)
+* [Trade-offs and Mitigations](#trade-offs-and-mitigations
+  * [Case-sensitivity](#case-sensitivity)
+  * [Security](#security)
+* [Prior Art and Alternatives](#prior-art-and-alternatives)
+  * [Alternatives and why they were not chosen](#alternatives-and-why-they-were-not-chosen)
+* [Open Questions](#open-questions)
+* [Future Possibilities](#future-possibilities)
+
+## Motivation
+
+The motivation for defining the specification for context and baggage
+propagation by using environment variables as carriers stems from a long open
+issue on the OpenTelemetry Specification repository, [issue #740](https://github.com/open-telemetry/opentelemetry-specification/issues/740#issue-665588273).
+
+The original issue identifies several use cases of systems that do not
+communicate across bounds by leveraging HTTP communications like, ETL, batch,
+and CI/CD systems, that need to benefit from context and baggage propagation.
+Defining a specification for this will enable better observability of these
+systems by enabling distributed tracing. There has already been a significant
+amount of [Prior Art](#prior-art) built within the industry and within
+OpenTelemetry to accomplish the immediate needs, however, OpenTelemetry at this
+time does not clearly define the specification for this form of propagation.
+
+Notably, as we define semantic conventions within the [CI/CD Working Group][cicd-wg],
+we'll need the specification defined for the industry to be able to adopt
+native tracing wtihin CI/CD systems.
+
+[cicd-wg]: https://github.com/open-telemetry/community/blob/main/projects/ci-cd.md
+
+## Explanation
+
+To propagate context and baggage between parent, and child processes
+in systems where HTTP communication does not occur, a specification for
+Environmental Variables can be used.
+
+Consider the following diagram:
+
+![Environment Variable Context Propagation](./img/0258-env-context-parent-child-process.png)
+
+In this diagram, a parent process is forked to spawn a child process,
+inheriting the environment variables from the original parent. The environment
+variables defined here, `TRACEPARENT`, `TRACESTATE`, and `BAGGAGE` are used to
+propagate context to the child process such that it can be tied to the parent.
+Without `TRACEPARENT`, a tracing backend would not be able to connect the child
```

Comment:
one thing that is unclear to me, are we proposing that the parent process should alter the env vars before creating a child process? It's never stated, the text reads like the child just inherits the same vars as parent. I think this would be a problem because if you imaging a hierarchy of child processes and they all inherit TRACEPARENT from the top parent process, then the hierarchy will not be represented in the trace (all spans will be children of the same root span). A common way to deal with this in OpenTelemetry is by having two levels of spans in each process, one for the entry point (`server` or `consumer` kinds) and another for each outbound IPC (`client` or `producer` spans). It's the latter that are propagated to downstream processes and form a call tree.

Reply:
So I think it would be useful if the spec outlined how instrumentation using the new env-var carrier should behave, i.e. if I create a "client" style span, how should its context be transmitted to the child process.

Reply:
That's a good point. I agree that there would be issues if they weren't able to alter the env vars and I agree that the spec should outline the behavior you mention. I can call out that specific need for the spec to define the behavior in this OTEP.

Reply:
🤔 when you pass TRACEPARENT you do it on purpose so your intention is that all the child spans have that parent context. If you do not want to pass that parnet context to all spans you should use OTEL_EXPORTER_OTLP_HEADERS or OTEL_EXPORTER_OTLP_TRACES_HEADERS, or OTEL_EXPORTER_OTLP_LOGS_HEADERS or OTEL_EXPORTER_OTLP_METRICS_HEADERS, or other implemented variables by the application used

Reply:
I think this has been addressed through refinement of the OTEP and gearing towards TextMap Propagation where  injector/extractor is defined, etc. Feel free to give another review and see if this change makes sense.